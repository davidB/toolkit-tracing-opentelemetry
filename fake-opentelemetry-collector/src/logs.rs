use crate::common::cnv_attributes;
use opentelemetry_proto::tonic::collector::logs::v1::{
    logs_service_server::LogsService, ExportLogsServiceRequest, ExportLogsServiceResponse,
};
use opentelemetry_proto::tonic::common::v1::AnyValue;
use serde::Serialize;
use std::collections::BTreeMap;
use std::sync::{mpsc, Mutex};

/// This is created to flatten the log record to make it more compatible with insta for testing
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct ExportedLog {
    pub trace_id: String,
    pub span_id: String,
    pub observed_time_unix_nano: u64,
    pub severity_number: i32,
    pub severity_text: String,
    pub body: Option<String>,
    pub attributes: BTreeMap<String, String>,
    pub dropped_attributes_count: u32,
    pub flags: u32,
}

impl From<opentelemetry_proto::tonic::logs::v1::LogRecord> for ExportedLog {
    fn from(value: opentelemetry_proto::tonic::logs::v1::LogRecord) -> Self {
        Self {
            trace_id: hex::encode(value.trace_id),
            span_id: hex::encode(value.span_id),
            observed_time_unix_nano: value.observed_time_unix_nano,
            severity_number: value.severity_number,
            severity_text: value.severity_text,
            body: value.body.map(|value| format!("{:?}", value)),
            attributes: cnv_attributes(&value.attributes),
            dropped_attributes_count: value.dropped_attributes_count,
            flags: value.flags,
        }
    }
}

pub(crate) struct FakeLogsService {
    tx: Mutex<mpsc::SyncSender<ExportedLog>>,
}

impl FakeLogsService {
    pub fn new(tx: mpsc::SyncSender<ExportedLog>) -> Self {
        Self { tx: Mutex::new(tx) }
    }
}

#[tonic::async_trait]
impl LogsService for FakeLogsService {
    async fn export(
        &self,
        request: tonic::Request<ExportLogsServiceRequest>,
    ) -> Result<tonic::Response<ExportLogsServiceResponse>, tonic::Status> {
        request
            .into_inner()
            .resource_logs
            .into_iter()
            .flat_map(|rl| rl.scope_logs)
            .flat_map(|sl| sl.log_records)
            .map(ExportedLog::from)
            .for_each(|el| self.tx.lock().unwrap().send(el).unwrap());
        Ok(tonic::Response::new(ExportLogsServiceResponse {
            partial_success: None,
        }))
    }
}
