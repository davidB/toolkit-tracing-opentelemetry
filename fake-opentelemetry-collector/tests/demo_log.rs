use std::time::Duration;

use fake_opentelemetry_collector::{setup_logger_provider, FakeCollectorServer};
use opentelemetry::logs::{LogRecord, Logger, LoggerProvider, Severity};
use tracing::debug;

#[tokio::test(flavor = "multi_thread")]
async fn demo_fake_logger_and_collector() {
    debug!("Start the fake collector");
    let mut fake_collector = FakeCollectorServer::start()
        .await
        .expect("fake collector setup and started");

    debug!("Init the 'application' & logger provider");
    let logger_provider = setup_logger_provider(&fake_collector).await;
    let logger = logger_provider.logger("test");

    debug!("Run the 'application' & send log ...");
    let mut record = logger.create_log_record();
    record.set_body("This is information".into());
    record.set_severity_number(Severity::Info);
    record.set_severity_text("info");
    logger.emit(record);

    debug!("Shutdown the 'application' & logger provider");
    let _ = logger_provider.force_flush();
    logger_provider
        .shutdown()
        .expect("no error during shutdown");
    drop(logger_provider);

    debug!("Collect & check the logs");
    let otel_logs = fake_collector
        .exported_logs(1, Duration::from_millis(500))
        .await;

    insta::assert_yaml_snapshot!(otel_logs, {
        "[].trace_id" => insta::dynamic_redaction(|value, _path| {
            assert2::let_assert!(Some(trace_id) = value.as_str());
            format!("[trace_id:lg{}]", trace_id.len())
        }),
        "[].span_id" => insta::dynamic_redaction(|value, _path| {
            assert2::let_assert!(Some(span_id) = value.as_str());
            format!("[span_id:lg{}]", span_id.len())
        }),
        "[].observed_time_unix_nano" => "[timestamp]",
        "[].severity_number" => 9,
        "[].severity_text" => "info",
        "[].body" => "AnyValue { value: Some(StringValue(\"This is information\")) }",
    });
}
