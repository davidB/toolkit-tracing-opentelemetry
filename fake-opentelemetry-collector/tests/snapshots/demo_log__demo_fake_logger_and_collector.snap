---
source: fake-opentelemetry-collector/tests/demo_log.rs
expression: otel_logs
snapshot_kind: text
---
- trace_id: "[trace_id:lg0]"
  span_id: "[span_id:lg0]"
  observed_time_unix_nano: "[timestamp]"
  severity_number: 9
  severity_text: info
  body: "AnyValue { value: Some(StringValue(\"This is information\")) }"
  attributes: {}
  dropped_attributes_count: 0
  flags: 0
